### PR TITLE
Add support of custom rules

### PR DESCRIPTION
**Pull request by VeselovAlex**
Added a way to pass custom rule scripts to linter.
* Use `-r` or `--rules` arg with path to js file following

### CODE REVIEW

**Comment by DavidAnson** [comment]
Wow, thank you!! Looking quickly on my phone, your changes make sense. But I will need to review more carefully this evening or weekend. One thing that would be nice to add before then are test cases. (And I’ll ask you to squash the changes into one.)

File: README.md [hunk 1]
```diff
@@ -24,6 +24,7 @@ $ markdownlint --help
     -s, --stdin                         read from STDIN (no files)
     -o, --output [outputFile]           write issues to file (no console)
     -c, --config [configFile]           configuration file (JSON or YAML)
+    -r, --rules  [file|directory|glob]  custom rule files
```
Comment (DavidAnson) [inline README.md]: Or packages? I think what you did for files will also support packages, so let’s say that here.

File: markdownlint.js [hunk 2]
```diff
@@ -41,14 +41,22 @@ function readConfiguration(args) {
   return config;
 }
 
-function prepareFileList(files) {
+function prepareFileList(files, fileExtensions) {
   var globOptions = {
     nodir: true
   };
+  var extensionGlobPart = '*';
+  if (fileExtensions && fileExtensions.length > 0) {
```
Comment (DavidAnson) [inline markdownlint.js]: This is an internal function, so is the parameter validation necessary?

File: markdownlint.js [hunk 3]
```diff
@@ -41,14 +41,22 @@ function readConfiguration(args) {
...
+    if (fileExtensions.length === 1) {
+      extensionGlobPart += '.' + fileExtensions[0];
+    } else {
+      extensionGlobPart += '.{' + fileExtensions.join(',') + '}';
```
Comment (DavidAnson) [inline markdownlint.js]: Will this syntax work for just one item? (I’d expect it to, but it needs testing.) If it does, the if/else can be removed.

File: markdownlint.js [hunk 4]
```diff
@@ -117,12 +125,23 @@ program
   .option('-s, --stdin', 'read from STDIN (no files)')
   .option('-o, --output [outputFile]', 'write issues to file (no console)')
   .option('-c, --config [configFile]', 'configuration file (JSON or YAML)')
+  .option('-r, --rules  [file|directory|glob]', 'custom rule files', concatArray, [])
```
Comment (DavidAnson) [inline markdownlint.js]: I like how you lined up the brackets here.

File: markdownlint.js [hunk 5]
```diff
@@ -117,12 +125,23 @@ program
...
 
 program.parse(process.argv);
 
-var files = prepareFileList(program.args);
-var ignores = prepareFileList(program.ignore);
+function loadCustomRuleFromFile(filepath) {
+  try {
+    return require(filepath.absolute);
+  } catch (err) {
+    console.error('Cannot load custom rule form ' + filepath.original + ': ' + err.message);
```
Comment (DavidAnson) [inline markdownlint.js]: form -> from

File: markdownlint.js [hunk 6]
```diff
@@ -117,12 +125,23 @@ program
...
+    process.exitCode = 3;
```
Comment (DavidAnson) [inline markdownlint.js]: I think we need to return or be sure not to run the code below some other way?

File: markdownlint.js [hunk 7]
```diff
@@ -117,12 +125,23 @@ program
...
+  }
+}
+
+var files = prepareFileList(program.args, ['md', 'markdown']);
+var ignores = prepareFileList(program.ignore, ['md', 'markdown']);
+var customRules = prepareFileList(program.rules, ['js']).map(loadCustomRuleFromFile);
```
Comment (DavidAnson) [inline markdownlint.js]: I like how you did this.

File: README.md (hunk 1 above)
Comment (DavidAnson) [reply README.md]: Wait, maybe that is wrong. The new code searches for all .js files, then require-s them all separately. So packages are not supported. What if it was used like this instead: —rules file.js dir/package

Comment (DavidAnson) [reply README.md]: Basically, just require each of the things that was passed to —rules.

**Comment by VeselovAlex** [comment]
`package.json` modifications:
* Test file directory declaration moved from scripts to `ava` parameters, ignoring `custom-rules` directory
* Add `test-rule-package` from `custom-rules` for testing

**Comment by DavidAnson** [comment]
Looks awesome! I’ll review in more detail in a day or two. Thanks so much!

File: markdownlint.js [hunk 8]
```diff
@@ -117,12 +124,37 @@ program
...
+function loadCustomRuleFromFile(filepath) {
+  try {
+    return require(filepath.absolute);
+  } catch (err) {
+    console.error('Cannot load custom rule from ' + filepath.original + ': ' + err.message);
+    process.exit(3);
+  }
+}
+
+function tryResolvePath(filepath) {
```
Comment (DavidAnson) [inline markdownlint.js]: Nice, I like this approach!

**Comment by DavidAnson** [comment]
Thank you again! I will publish a new version in a week or so.

**Pull request merged**